src/functional/otf/workflow.py

```diff
@@ -0,0 +1,129 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import dataclasses
+from typing import Generic, Protocol, TypeVar
+
+
+StartT = TypeVar("StartT")
+StartT_contra = TypeVar("StartT_contra", contravariant=True)
+EndT = TypeVar("EndT")
+EndT_co = TypeVar("EndT_co", covariant=True)
+NewEndT = TypeVar("NewEndT")
+IntermediateT = TypeVar("IntermediateT")
+
+
+def make_step(function: StepProtocol[StartT, EndT]) -> Step[StartT, EndT]:
+    """
+    Wrap a function in the workflow step convenience wrapper.
+
+    Examples:
+    ---------
+    >>> @make_step
+    ... def times_two(x: int) -> int:
+    ...    return x * 2
+
+    >>> def stringify(x: int) -> str:
+    ...    return str(x)
+
+    >>> # create a workflow int -> int -> str
+    >>> times_two.chain(stringify)(3)
+    '6'
+    """
+    return Step(function)
+
+
+class StepProtocol(Protocol[StartT_contra, EndT_co]):
+    """
+    Workflow step protocol.
+
+    Anything that implements this interface can be a workflow step.
+    - callable
+    - take a single input argument
+    """
+
+    def __call__(self, inp: StartT_contra) -> EndT_co:
+        ...
+
+
+@dataclasses.dataclass(frozen=True)
+class Step(Generic[StartT, EndT]):
+    """
+    Workflow step convenience wrapper.
+
+    Can wrap any callable which implements the workflow step protocol,
+    adds the .chain(other_step) method for convenience.
+
+    Examples:
+    ---------
+    >>> def times_two(x: int) -> int:
+    ...    return x * 2
+
+    >>> def stringify(x: int) -> str:
+    ...    return str(x)
+
+    >>> # create a workflow int -> int -> str
+    >>> Step(times_two).chain(stringify)(3)
+    '6'
+    """
+
+    step: StepProtocol[StartT, EndT]
+
+    def __call__(self, inp: StartT) -> EndT:
+        return self.step(inp)
+
+    def chain(self, step: StepProtocol[EndT, NewEndT]) -> Workflow[StartT, EndT, NewEndT]:
+        return Workflow(first=self.step, second=step)
+
+
+@dataclasses.dataclass(frozen=True)
+class Workflow(Generic[StartT, IntermediateT, EndT]):
```

Comment:
As I mentioned in person, you could rename this to `CombinedStep` or something to emphasize the use of `Step.chain` like you recommended as opposed to manually using `Workflow`.

Reply:
Sure, but just to check first: In a few places I have placed `Final[Workflow[...]]` annotations to make it obvious that a variable is a Workflow going from this stage to that. Are you willing to make the readability sacrifice that comes with changing these annotations to `Final[CombinedStep[...]]`?

Reply:
I'm not such a pro on these type annotations, but can you not use `StepProtocol` at those places so that it accepts any step, which `Workflow` is?

Reply:
Is the readability delta of `Workflow` vs `CombinedStep` really any different from the one of `Workflow` vs `StepProtocol`? As in, I am trying to convey that this is a *workflow*. Is telling the user that it is a StepProtocol, which is technically equivalent to a workflow going to achieve that goal?